File: programmers/python/programmers_bruteforce_2.py
sieve = [True] * 10000001
check = [False] * 10000001
num_check = [False] * 8
answer = 0

def Eratos():
	n = 10000001
	m = int(n ** 0.5)
	for i in range(2, m + 1):
		if sieve[i] == True:
			for j in range(i + i, n, i):
				sieve[j] = False
	sieve[0] = False
	sieve[1] = False
	return
  
def dfs(idx, n, num, numbers):
	if sieve[num] == True and check[num] == False:
		global answer
		answer += 1
		check[num] = True
	if idx >= n:
		return
	for i in range(n):
		if num_check[i] == False:
			num_check[i] = True
			dfs(idx + 1, n, int(str(num) + numbers[i]), numbers)
			num_check[i] = False
	return
	
def solution(numbers):
	Eratos()
	dfs(0, len(numbers), 0, numbers)
	return answer

File: programmers/cpp/programmers_bruteforce_2.cpp
#include <string>
#include <vector>

using namespace std;

int mem[10000000];
int check[10000000];
int flag[7];
int answer;

void prime_num()
{
	for (int i = 2; i < 10000000; i++)
		mem[i] = 1;
	for (int i = 2; i * i < 10000000; i++)
	{
		if (mem[i] == 1)
		{
			for (int j = i * i; j <= 10000000; j += i)
				mem[j] = 0;
		}
	}
}

void dfs(string num, string numbers)
{
	if (num.size() > numbers.size() || num[0] == '0')
		return;
	if (mem[atoi(num.c_str())] == 1)
	{
		if (check[atoi(num.c_str())] == 0)
		{
			answer++;
			check[atoi(num.c_str())] = 1;
		}
	}
	for (int i = 0; i < numbers.size(); i++)
	{
		if (flag[i] == 0)
		{
			flag[i] = 1;
			dfs(num + numbers[i], numbers);
			flag[i] = 0;
		}
	}
}

int solution(string numbers) {
	prime_num();
	dfs("", numbers);
	return answer;
}